test/lib.bash

```diff
@@ -115,12 +115,131 @@ function downstream_eventing_e2e_tests {
   return $failed
 }
 
+# == Upgrade testing
+
+function run_rolling_upgrade_tests {
+  logger.info "Running rolling upgrade tests"
+  (
+  local latest_cluster_version latest_serving_version latest_eventing_version \
+    rootdir scope serving_in_scope eventing_in_scope serving_prober_pid \
+    eventing_prober_pid prev_serving_version prev_eventing_version \
+    ocp_target_version
+
+  scope="${1:?Provide an upgrade scope as arg[1]}"
+  serving_in_scope="$(echo "${scope}" | grep -vq serving ; echo "$?")"
+  eventing_in_scope="$(echo "${scope}" | grep -vq eventing ; echo "$?")"
+
+  prev_serving_version="$(actual_serving_version)"
+  prev_eventing_version="$(actual_eventing_version)"
+
+  # Save the rootdir before changing dir
+  rootdir="$(dirname "$(dirname "$(realpath "${BASH_SOURCE[0]}")")")"
+
+  (( serving_in_scope )) && prepare_knative_serving_tests
+
+  logger.info 'Testing with pre upgrade tests'
+
+  (( serving_in_scope )) && run_serving_preupgrade_test
+  (( eventing_in_scope )) && run_eventing_preupgrade_test
+
+  logger.info 'Starting prober tests'
+
+  if (( serving_in_scope )); then
+    start_serving_prober "${prev_serving_version}" /tmp/prober-pid
+    serving_prober_pid=$(cat /tmp/prober-pid)
+  fi
+  if (( eventing_in_scope )); then
+    start_eventing_prober /tmp/prober-pid
+    eventing_prober_pid=$(cat /tmp/prober-pid)
+  fi
+
+  (( serving_in_scope )) && wait_for_serving_prober_ready
+  (( eventing_in_scope )) && wait_for_eventing_prober_ready
+
+  if [[ $UPGRADE_SERVERLESS == true ]]; then
+    latest_serving_version="${KNATIVE_SERVING_VERSION/v/}"
+    latest_eventing_version="${KNATIVE_EVENTING_VERSION/v/}"
+
+    logger.info "Updating Serverless to ${CURRENT_CSV}"
+    logger.debug "Serving version: ${prev_serving_version} -> ${latest_serving_version}"
+    logger.debug "Eventing version: ${prev_eventing_version} -> ${latest_eventing_version}"
+
+    approve_csv "$CURRENT_CSV" "$OLM_UPGRADE_CHANNEL"
+    (( serving_in_scope )) && check_serving_upgraded "${latest_serving_version}"
+    (( eventing_in_scope )) && check_eventing_upgraded "${latest_eventing_version}"
+  fi
+
+  # Might not work in OpenShift CI but we want it here so that we can consume
+  # this script later and re-use
+  if [[ $UPGRADE_CLUSTER == true ]]; then
+    # End the prober test now before we start cluster upgrade, up until now we
+    # should have zero failed requests. Cluster upgrade will fail probers as
+    # stuff is moved around.
+    (( serving_in_scope )) && end_serving_prober "${serving_prober_pid}"
+    (( eventing_in_scope )) && end_eventing_prober "${eventing_prober_pid}"
+
+    upgrade_ocp_cluster "${UPGRADE_OCP_IMAGE:-}"
+  fi
+
+  (( serving_in_scope )) && wait_for_serving_test_services_settle
+
+  logger.info "Running postupgrade tests"
+
+  (( serving_in_scope )) && run_serving_postupgrade_test
+  (( eventing_in_scope )) && run_eventing_postupgrade_test
+
+  (( serving_in_scope )) && end_serving_prober "${serving_prober_pid}"
```

Comment:
I'm curious if this will work if we already called this function before upgrading cluster. I.e. is it idempotent?

Reply:
It should be.